### DISCUSSION TITLE
Accessing resources in an iOS bundle (beta)

### DISCUSSION THREAD

**Discussion by stijnfrishert** [tauri-apps/tauri#9810] [Q&A]
Hello everyone,

I'm building a mobile app for iOS, and I'd like to package some resource files along with the bundle. I want to read the (audio) files in within Rust, but send their relative paths over from the frontend.

All these files live in `../../../assets` (or subfolders therein), relative to `tauri.config.json`.

I've set up `["bundle"]["resources"] = ["../../../assets"]` in `tauri.config.json`, but this doesn't seem to be enough. The v1 guide mentions an `allowList`, but I can't seem to find anything about that in the sparse v2 docs.

I'm also not sure whether specifying it like this will copy the `assets` folder, or it's children, and whether it'll keep the internal folder structure.

In other words: how do I put external files in my bundle, so that I can send the path through a command and load them from my rust code _with the v2 beta?_

---

Just found out I can view the web console when right clicking the window, and when trying to access the file with `resolveResource()` it does give me a 400 bad request error.

`Unhandled Promise Rejection: path.resolve_directory not allowed. Permissions associated with this command: path:allow-resolve-directory, path:default`

---

Found a v2 guide about [permissions](https://v2.tauri.app/security/permissions/), but I'm still a bit baffled how I'm supposed to tie this all together 😅

**Comment by FabianLars** [comment]
> Found a v2 guide about [permissions](https://v2.tauri.app/security/permissions/), but I'm still a bit baffled how I'm supposed to tie this all together 😅

There should be a json file in `src-tauri/capabilities/` already. Open that and you should see a `permissions` array. Simply add the missing permissions to that and the error should go away (not sure if it will work then, can't remember whether resources are implemented on mobile).

**Reply by stijnfrishert** [reply]
There's no `src-tauri/capabilities` folder at all. I just generated a temporary new mobile project using `cargo create-tauri-app --beta` and that too doesn't contain any capabitilies (is that a bug I should report/create an issue for?).

I'm happy to add the missing permissions, but what file should I start from? Is that the first code snippet on [the capabilities page](https://beta.tauri.app/security/capabilities/) minus `window::allow-set-title`?

```json
{
  "$schema": "./schemas/desktop-schema.json",
  "identifier": "main-capability",
  "description": "Capability for the main window",
  "windows": ["main"],
  "permissions": [
    "path:default",
    "event:default",
    "window:default",
    "app:default",
    "resources:default",
    "menu:default",
    "tray:default",
    "window:allow-set-title"
  ]
}
```

**Reply by stijnfrishert** [reply]
I've added `src-tauri/capabilities/main.json5` with these contents, but I get the same error in my console:

```json5
{
  $schema: "./schemas/desktop-schema.json",
  identifier: "main-capability",
  description: "Capability for the main window",
  windows: ["main"],
  permissions: [
    "path:default",
    "path:allow-resolve-directory",
    "event:default",
    "window:default",
    "app:default",
    "resources:default",
    "menu:default",
    "tray:default",
    // "window:allow-set-title",
  ],
}
```

> `Unhandled Promise Rejection: path.resolve_directory not allowed. Permissions associated with this command: path:allow-resolve-directory, path:default`

**Reply by FabianLars** [reply]
Can you check the version of your create-tauri-app tool?

**Reply by stijnfrishert** [reply]
@FabianLars `3.13.17`

Installed it this week.

**Reply by FabianLars** [reply]
Hmm that's weird then. I'm not at home right now so I can't check it out myself, therefore I would take you up on the offer to open an issue, thank you :)

**Reply by FabianLars** [reply]
And for the still occurring error, can you post the output of the tauri info command to check if CTA messed up something else?

**Reply by stijnfrishert** [reply]
> Hmm that's weird then. I'm not at home right now so I can't check it out myself, therefore I would take you up on the offer to open an issue, thank you :)

Here you are! :) https://github.com/tauri-apps/tauri/issues/9814

> And for the still occurring error, can you post the output of the tauri info command to check if CTA messed up something else?

`cargo tauri info` (with the tauri project as cwd) gives me:

```
[✔] Environment
    - OS: Mac OS 14.4.1 X64
    ✔ Xcode Command Line Tools: installed
    ✔ rustc: 1.78.0 (9b00956e5 2024-04-29)
    ✔ cargo: 1.78.0 (54d8815d0 2024-03-26)
    ✔ rustup: 1.27.1 (54dd3d00f 2024-04-24)
    ✔ Rust toolchain: stable-x86_64-apple-darwin (environment override by RUSTUP_TOOLCHAIN)
    - node: 20.12.2
    - pnpm: 9.0.6
    - npm: 10.5.0

[-] Packages
    - tauri [RUST]: 2.0.0-beta.19
    - tauri-build [RUST]: 2.0.0-beta.15
    - wry [RUST]: 0.39.5
    - tao [RUST]: 0.28.0
    - tauri-cli [RUST]: 2.0.0-beta.17
    - @tauri-apps/api [NPM]: 2.0.0-beta.11
    - @tauri-apps/cli [NPM]: 2.0.0-beta.17

[-] App
    - build-type: bundle
    - CSP: unset
    - frontendDist: ../dist
    - devUrl: http://localhost:1420/
    - framework: React
    - bundler: Vite

[-] iOS
    - Developer Teams: <redacted>
```

**Reply by stijnfrishert** [reply]
@FabianLars I managed to silence the warnings by adding this `capabilities/main.json`

```json
{
  "identifier": "main-capability",
  "description": "Capability for the main window",
  "windows": ["main"],
  "permissions": [
    "path:default",
    {
      "identifier": "path:allow-resolve-directory",
      "allow": [{ "path": "$RESOURCE" }]
    }
  ]
}
```

However, when I log the path that comes out of `resolveResource('sounds/correct.wav')`, I get `/Users/<name>/Development/mimi2/target/x86_64-apple-darwin/debug/sounds/correct.wav`. When I check the debug build folder, none of the assets seem to have been copied over/linked. Is that something I should do manually?

I'm building an iOS app here, but testing on macOS.

Any pointers in the right direction are appreciated :)

**Reply by lyager** [reply]
@stijnfrishert Did you manage to get further on this? I'm trying to add binary files to my IOS app, but when calling `resourceDir()` from typescript, I'm just getting `unknown path`. I have a capabilities file in `src-tauri/capabilities/` but it was generated some time ago when I started out.

**Reply by lyager** [reply]
AFAIKS these are not implemented for IOS yet?

**Reply by stijnfrishert** [reply]
@lyager Not yet. I ran into other issues building Tauri locally. Those have been resolved, next I need to finish off https://github.com/tauri-apps/tauri/pull/9992, and after that I hope to take a look at assets. I need both the aforementioned PR and assets for my mobile app, so 🤞

**Reply by lyager** [reply]
@stijnfrishert Thanks for the reply - I created this https://github.com/tauri-apps/tauri/issues/10070 and I have a workaround in my local Tauri repo.

**Comment by stijnfrishert** [comment]
Ok, took me a while to figure out what was wrong here (needed [frameworks for iOS](https://github.com/tauri-apps/tauri/issues/9962) first).

Long stor short, the issue here seems to be that the binary resource system simply hasn't been implemented for iOS yet.

There's [more](https://github.com/tauri-apps/tauri/issues/10070) [people](https://github.com/tauri-apps/tauri/issues/10645) running into that though, so I suppose feature parity on this front with regular macOS builds is something people expect to work.

Is this something that's on the roadmap for the initial v2 release?

**Reply by FabianLars** [reply]
> Is this something that's on the roadmap for the initial v2 release?

No, 2.0.0 is pretty much in feature freeze right now. I expect this to be high-ish in prio for one of the first 2.x minor releases though.
